### ISSUE TITLE
Add Initial PHP 8 Support (PDO, Date, Comment, SimpleTest).

### ISSUE THREAD

**Issue by hosef** [backdrop/backdrop-issues#4772]
**Description of the need**

PHP 8 will be out soon, and we will need to make some changes so that Backdrop can run on it. 

**Additional information**

There is a meta issue for adding PHP 8 support for Drupal 9, but I was not able to find information about PHP 8 support in older versions of Drupal.
https://www.drupal.org/project/drupal/issues/3109885

---

PR by @hosef https://github.com/backdrop/backdrop/pull/3412

**Comment by olafgrabienski** [comment]
Here’s the issue about D7: https://www.drupal.org/project/drupal/issues/3145797

**Comment by hosef** [comment]
The change that I think will be most impactful is that the `DatabaseConnection` class can no longer extend the `PDO` class because fatal errors are now thrown when methods on abstract classes do not match the signature of the function they are overriding or implementing. For Drupal 9 , they are changing the `DatabaseConnection` class to wrap an instance of the `PDO` class instead of extending it. I have made a similar change here, because that seems like it will be the smallest change to allow us to 

This may break backwards compatibility for modules that depend on some of the more complex usages of the `PDO` class.

**Comment by klonos** [comment]
I would like to test this, but will have to wait till Lando supports php8 versions: https://docs.lando.dev/config/php.html#supported-versions

**Comment by indigoxela** [comment]
> This may break backwards compatibility for modules that depend on some of the more complex usages of the PDO class.

Is there an actual example for such usages?

**Comment by hosef** [comment]
I can't think of anything off the top of my head. If you are just using db_query() or other functions like that you will be fine. I just wanted people to be aware in case they were doing something weird.

**Comment by indigoxela** [comment]
> I just wanted people to be aware in case they were doing something weird.

CiviCRM might be such a candidate, not sure.

Anyway, I'm planning to install PHP8 on my dev server soon, so I will be able to help out with testing.

**Comment by indigoxela** [comment]
> I'm planning to install PHP8 on my dev server soon

Done. Yes, this will need some work, WSOD with the second installer step. :grin:

**Comment by hosef** [comment]
Does it WSOD when you use the branch from this PR?

**Comment by indigoxela** [comment]
> Does it WSOD when you use the branch from this PR?

@hosef sorry, I missed your PR, just tested it now. It installs just fine. No errors, nothing in dblog.

One thing I noticed: I added another post without problems, but that post doesn't show up in the content overview (admin/content). I suspect, something with views db handling might need an update, too.

**Comment by indigoxela** [comment]
To verify I created a new view (without exposed filters) and in that one all new content shows up. Then I added the exposed filters one by one to figure out which one causes the problem. It's "Content: Has new content (exposed)". As soon as I add that, the new items disappear. The two one that get created on install are still there.

Happy digging. :wink:

**Comment by jlfranklin** [comment]
> Is there an actual example for such usages?

[Database SSL Check](https://www.drupal.org/project/database_ssl_check).   I suppose it could be patched to check if the database connect is-a PDO object or has-a PDO object.

**Comment by klonos** [comment]
> I would like to test this, but will have to wait till Lando supports php8 versions: https://docs.lando.dev/config/php.html#supported-versions

Well, it does now 😄 ...I just tested installing a vanilla Backdrop 1.x on php8, and I got this error when trying to launch the site:

```
Fatal error: Declaration of DatabaseConnection::query($query, array $args = [], $options = [])
must be compatible with PDO::query(string $query, ?int $fetchMode = null, mixed ...$fetchModeArgs)
in /app/docroot/core/includes/database/database.inc on line 676
```

Trying the same with the PR changes applied still throws some errors (lando-related, NOT Backdrop), but I at least got install.php to load and to install the site w/o issues.

Drush may also need fixing, as the scripts that I use regularly with php 5.6 and 7.x fail, with this error being repeated:

```
OCI runtime exec failed: exec failed: container_linux.go:370: starting container process caused:
exec: "drush": executable file not found in $PATH: unknown
```

**Comment by klonos** [comment]
PS: for anyone else using lando, the drush issues seem to be related to composer, and it seems to be fixed with https://github.com/lando/lando/pull/2729 ...this change is included in https://github.com/lando/lando/releases/tag/v3.0.25 but that's a pre-release, and I've installed using https://formulae.brew.sh/cask/lando which is still in v3.0.23. I'll test this again once there's an official lando release.

**Comment by quicksketch** [comment]
I tried out the PR at https://github.com/backdrop/backdrop/pull/3412 and found it to help substantially with PHP 8 issues. The changed code all seems good to me, correlating to various Drupal 7 fixes.

I tried to run the full test suite and ran into several more issues, which I fixed by adding an additional commit that includes these issues:

- https://www.drupal.org/project/drupal/issues/3200708
- https://www.drupal.org/project/drupal/issues/3206429
- https://www.drupal.org/project/drupal/issues/3210215

However, there are still more to be done, and some of these issues (such as https://www.drupal.org/project/drupal/issues/3204161) are so large that perhaps they should be done in individual pull requests rather than all together like this.

I'm leaning towards merging this initial PR to get the current fixes in, then iterate on it to solve the more esoteric issues.

**Comment by klonos** [comment]
I agree with the above ^^ ...baby steps. Some of these commits I was planning to get into separate PRs (so that they can be more easily, individually reverted if need be), as a follow-up to #5042.

**Comment by quicksketch** [comment]
Thanks for your review @klonos, I made the adjustments you suggested so the PR could use another review: https://github.com/backdrop/backdrop/pull/3412

**Comment by klonos** [comment]
Thanks @quicksketch ...I've added a suggestion to add an inline comment, to explain things for posterity (so that people don't wonder the same thing as I) + I've created this PR, which includes changes in `modules/simpletest/simpletest.test` from https://git.drupalcode.org/project/drupal/commit/4c11b77 which seem to have been missed: https://github.com/hosef/backdrop/pull/1/files

**Comment by quicksketch** [comment]
I merged in @klonos's PR to include the missing portions of the D7 patch. All tests should now be passing.

We discussed this during the last weekly call and no one objected to the "merge now, iterate later" approach for this initial PHP 8 support PR. So I'm removing the "needs work" and I think this initial work is ready to go.

**Comment by quicksketch** [comment]
I have merged https://github.com/backdrop/backdrop/pull/3412 into 1.x. As this would make for a good 1.19.0 promotional feature, and the likelihood of regressions is higher than a normal bug fix, I would prefer not to merge this into 1.18.x.

We still have more work to do here, I filed https://github.com/backdrop/backdrop-issues/issues/5076 as a full meta so we can start tracking follow-ups.
